pkg/datastore/pagination/iterator.go

```diff
@@ -0,0 +1,114 @@
+package pagination
+
+import (
+	"context"
+	"errors"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/datastore/options"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// NewPaginatedIterator creates an implementation of the datastore.Iterator
+// interface that internally paginates over datastore results.
+func NewPaginatedIterator(
+	ctx context.Context,
+	reader datastore.Reader,
+	filter datastore.RelationshipsFilter,
+	pageSize uint64,
+	order options.SortOrder,
+) (datastore.RelationshipIterator, error) {
+	pi := &paginatedIterator{
+		ctx:      ctx,
+		reader:   reader,
+		filter:   filter,
+		pageSize: pageSize,
+		order:    order,
+		delegate: common.NewSliceRelationshipIterator(nil, options.ByResource),
+	}
+
+	pi.startNewBatch(nil)
+
+	return pi, pi.err
+}
+
+type paginatedIterator struct {
+	ctx      context.Context
+	reader   datastore.Reader
+	filter   datastore.RelationshipsFilter
+	pageSize uint64
+	order    options.SortOrder
+
+	delegate          datastore.RelationshipIterator
+	returnedFromBatch uint64
+	err               error
+	closed            bool
+}
+
+func (pi *paginatedIterator) Next() *core.RelationTuple {
+	if pi.Err() != nil {
+		return nil
+	}
+
+	var next *core.RelationTuple
+	for next = pi.delegate.Next(); next == nil; next = pi.delegate.Next() {
+		if pi.delegate.Err() != nil {
+			pi.err = pi.delegate.Err()
+			return nil
+		}
```

Comment:
nit: 🤔 handing of error state seems inconsistent in this method. What do you think about adding an additional case to `Err()` and checking here `pi.Err()`, and also below after `startNewBatch`?

```suggestion
		if pi.Err() != nil {
			return nil
		}
```

Reply:
Why? I think as written it's more clear when and where a particular error is going to occur and get copied and become sticky in the parent iterator. For example, if we just checked delegate error in the err block, we could accidentally overwrite delegate and clear the error state.